docs/design/rubygems.md

```diff
@@ -0,0 +1,344 @@
+      bundler (>= 1.15.0)
+      railties (= 6.1.4)
+      sprockets-rails (>= 2.0.0)
+...
+```
+
+#### Git dependencies
+
+Example of a git dependency in the `Gemfile.lock` file:
+
+```Gemfile.lock
+...
+GIT
+  remote: https://github.com/porta.git
+  revision: 779beabd653afcd03c4468e0a69dc043f3bbb748
+  branch: main
+  specs:
+    porta (2.14.1)
+...
+```
+
+Bundler uses this [format](https://github.com/rubygems/rubygems/blob/3da9b1dda0824d1d770780352bb1d3f287cb2df5/bundler/lib/bundler/source/git.rb#L130) to cache git repositories:
+
+```ruby
+"#{base_name}-#{shortref_for_path(revision)}"
+```
+
+Any other format will cause bundler to re-download the repository -> cache invalidation -> the build will fail.
+
+#### Path dependencies
+
+Example of a path dependency in the `Gemfile.lock` file:
+
+```Gemfile.lock
+...
+PATH
+  remote: ../foobar
+  specs:
+    foobar (0.1.0)
+...
+```
+
+All path dependencies must be in the project directory.
+Bundler [does not copy](https://github.com/rubygems/rubygems/blob/master/bundler/lib/bundler/source/path.rb#L83)
+those dependencies that are already within the root directory of the project.
+
+#### Plugins
+
+Not supported by cachi2.
+
+### Checksums
+
+Checksum validation is enabled by default.
+It can be disabled with the `BUNDLE_DISABLE_CHECKSUM_VALIDATION` environment variable.
+
+There is also an option to generate checksums in `Gemfile.lock`, but in very weird way.
+(At least, I have not found any other way to do it.)
+
+```shell
+# manually add `CHECKSUMS` section somewhere to the Gemfile.lock
+vim Gemfile.lock
+# install any gem
+bundle add rails --version "6.1.7"
+# check the Gemfile.lock /o\
+cat Gemfile.lock
+```
+
+Example of a checksum section in the `Gemfile.lock` file from my custom [repository](https://github.com/slimreaper35/cachi2-rubygems.git):
+
+```Gemfile.lock
+...
+DEPENDENCIES
+  rails (= 6.1.7)
+
+CHECKSUMS
+  actioncable (6.1.7) sha256=ee5345e1ac0a9ec24af8d21d46d6e8d85dd76b28b14ab60929c2da3e7d5bfe64
+  actionmailbox (6.1.7) sha256=c4364381e724b39eee3381e6eb3fdc80f121ac9a53dea3fd9ef687a9040b8a08
+  actionmailer (6.1.7) sha256=5561c298a13e6d43eb71098be366f59be51470358e6e6e49ebaaf43502906fa4
+  actionpack (6.1.7) sha256=3a8580e3721757371328906f953b332d5c95bd56a1e4f344b3fee5d55dc1cf37
+  actiontext (6.1.7) sha256=c5d3af4168619923d0ff661207215face3e03f7a04c083b5d347f190f639798e
+  actionview (6.1.7) sha256=c166e890d2933ffbb6eb2a2eac1b54f03890e33b8b7269503af848db88afc8d4
+  ...
+
+BUNDLED WITH
+   2.5.14
+```
+
+I believe this feature is available since bundler [v2.5.0](https://github.com/rubygems/rubygems/blob/master/bundler/lib/bundler/lockfile_parser.rb#L55)
+from this [PR](https://github.com/rubygems/rubygems/pull/6374) being merged on Oct 21, 2023.
+
+### Environment variables
+
+The order of precedence for bundler configuration options is as follows:
+
+1. Local config (`<project_root>/.bundle/config or $BUNDLE_APP_CONFIG/config`)
+2. Environmental variables (ENV)
```

Comment:
Damn, local config takes precedence over ENV vars? That means if we just set the env vars, we may fail to configure the build for offline installation (if the user has some conflicting settings in their config). We might want overwrite their config instead (via `inject-files`) or parse their config and raise an error if we detect conflicting settings

Reply:
Only BUNDLE_APP_CONFIG takes precedence

Reply:
There is also BUNDLE_IGNORE_CONFIG, but this will ignore _all_ environment variables

Reply:
> There is also BUNDLE_IGNORE_CONFIG, but this will ignore all config variables

Yeah that doesn't sound like a good option, we don't want to mess with the users' settings that much

> Only BUNDLE_APP_CONFIG takes precedence

so we could potentially do something like
* copy the .bundle/config from the repo into the cachi2 output dir
* set the variables we need in the copied config
* set `BUNDLE_APP_CONFIG=<configfile in output dir>`

Although simply overwriting their config via `inject-files` would be fine too, copying it elsewhere is just an optional nice-to-have improvement IMO

Reply:
How could we set BUNDLE_CACHE_PATH in our copied config file before we know ${output_dir} ?